src/btree/bt_read.c

```diff
@@ -84,6 +84,119 @@ __evict_force_check(WT_SESSION_IMPL *session, WT_REF *ref)
+        /*
+         * We apply deltas from newest to oldest, ignore keys that have already got a delta update.
+         */
+        if (cbt.compare == 0) {
+            if (cbt.ins != NULL) {
+                if (cbt.ins->upd != NULL && F_ISSET(cbt.ins->upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            } else {
+                rip = &page->pg_row[cbt.slot];
+                first_upd = WT_ROW_UPDATE(page, rip);
+                if (first_upd != NULL && F_ISSET(first_upd, WT_UPDATE_RESTORED_FROM_DELTA))
+                    continue;
+            }
+        }
+
+        if (F_ISSET(&unpack, WT_DELTA_IS_DELETE)) {
+            WT_RET(__wt_upd_alloc_tombstone(session, &tombstone, &tmp_size));
+            F_SET(tombstone, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+            size += tmp_size;
+            upd = tombstone;
+        } else {
+            value.data = unpack.value;
+            value.size = unpack.value_size;
+            WT_RET(__wt_upd_alloc(session, &value, WT_UPDATE_STANDARD, &standard_value, &tmp_size));
+            standard_value->start_ts = unpack.tw.start_ts;
+            standard_value->durable_ts = unpack.tw.durable_start_ts;
+            F_SET(standard_value, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+            size += tmp_size;
+
+            if (WT_TIME_WINDOW_HAS_STOP(&unpack.tw)) {
+                WT_RET(__wt_upd_alloc_tombstone(session, &tombstone, &tmp_size));
+                tombstone->start_ts = unpack.tw.stop_ts;
+                tombstone->durable_ts = unpack.tw.durable_stop_ts;
+                F_SET(tombstone, WT_UPDATE_DURABLE | WT_UPDATE_RESTORED_FROM_DELTA);
+                tombstone->next = standard_value;
+                upd = tombstone;
+            } else
+                upd = standard_value;
+        }
+
+        WT_ERR(__wt_row_modify(&cbt, &key, NULL, &upd, WT_UPDATE_INVALID, true, true));
+
+        total_size += size;
+    }
+    WT_CELL_FOREACH_END;
+
+    /*
+     * The data is written to the disk so we can mark the page clean after re-instantiating prepared
+     * updates to avoid reconciling the page every time.
+     */
+    __wt_page_modify_clear(session, page);
+    __wt_cache_page_inmem_incr(session, page, total_size);
+
+    if (0) {
+err:
+        if (standard_value != NULL)
+            __wt_free(session, standard_value);
+        if (tombstone != NULL)
+            __wt_free(session, tombstone);
+    }
+    WT_TRET(__wt_btcur_close(&cbt, true));
+    return (ret);
+}
+
+/*
+ * __bt_reconstruct_deltas --
+ *     Reconstruct deltas on a page
+ */
+static int
+__bt_reconstruct_deltas(WT_SESSION_IMPL *session, WT_REF *ref, WT_ITEM *deltas, size_t delta_size)
+{
+    int i;
+
+    /* TODO: this is not the optimal algorithm. We can optimize this by using a min heap. */
+    for (i = (int)delta_size - 1; i >= 0; --i)
+        WT_RET(__bt_reconstruct_delta(session, ref, &deltas[i]));
```

Comment:
I know you have a comment elsewhere, but probably good to have a comment here explaining why reverse order.